lib/ansible/module_utils/common/file.py

```diff
@@ -115,88 +117,82 @@ class LockTimeout(Exception):
-    '''
-    def __init__(self):
-        self.lockfd = None
-
-    @contextmanager
-    def lock_file(self, path, tmpdir, lock_timeout=None):
-        '''
-        Context for lock acquisition
-        '''
-        try:
-            self.set_lock(path, tmpdir, lock_timeout)
-            yield
-        finally:
-            self.unlock()
-
-    def set_lock(self, path, tmpdir, lock_timeout=None):
-        '''
-        Create a lock file based on path with flock to prevent other processes
-        using given path.
-        Please note that currently file locking only works when it's executed by
-        the same user, I.E single user scenarios
-
-        :kw path: Path (file) to lock
-        :kw tmpdir: Path where to place the temporary .lock file
-        :kw lock_timeout:
-            Wait n seconds for lock acquisition, fail if timeout is reached.
-            0 = Do not wait, fail if lock cannot be acquired immediately,
-            Default is None, wait indefinitely until lock is released.
-        :returns: True
-        '''
-        lock_path = os.path.join(tmpdir, 'ansible-{0}.lock'.format(os.path.basename(path)))
-        l_wait = 0.1
-        r_exception = IOError
-        if sys.version_info[0] == 3:
-            r_exception = BlockingIOError
-
-        self.lockfd = open(lock_path, 'w')
-
-        if lock_timeout <= 0:
-            fcntl.flock(self.lockfd, fcntl.LOCK_EX | fcntl.LOCK_NB)
-            os.chmod(lock_path, stat.S_IWRITE | stat.S_IREAD)
+    """ Class for file locking, to avoid race conditions between multiple ansible processes.
+        Keep in mind this is advisory locking only (ansible's internal agreement). Any other process that doesn't
+        respect or know about our locking strategy can still read/write/remove original file.
+        In Python fcntl.lockf function should be used, which actually calls fcntl system API.
+        You can read more at https://apenwarr.ca/log/20101213
+    """
+    def __init__(self, path, lock_timeout=15, check_mode=False):
+        self.lock_directory = None
+        self.lock_file = None
+        self.path = path
+        self.lock_timeout = lock_timeout
+        self.check_mode = check_mode
+
+    def __enter__(self):
+        """" Set lock
+             We are setting lock on separate .lock file. Cannot lock original file, as ansible modules usually use
+             atomic_move function that changes inode, on which lock actually relies.
+         """
+        if self.check_mode:
             return True
-
-        if lock_timeout:
-            e_secs = 0
-            while e_secs < lock_timeout:
+        else:
+            # The default umask is 0o22 which turns off write permission of group and others
+            os.umask(0)
+
+            # Assure directory for .lock files
+            # On linux this is /tmp/
+            # On macos /var/folders/**/***/T/ which only has current user permissions (multi-user locking won't work)
+            sys_temp = tempfile.gettempdir()
```

Comment:
this won't always give the same result, it depends on env vars, that can vary by user and even task .. or task loop:
```
loop: [1,2,3]
envioronment: '{{ difftempvar[item]}}'
```

Reply:
Thanks for pointing this out. Hmm...any suggestion?
I tried to avoid hardcoding temp dir to '/tmp', but i don't see any other solution at the moment.

Other option i considered was placing .lock files beside original files. But this adds complexity...how to remove these files.

Reply:
even hardcoding /tmp is not a good option, some systems create a per user /tmp mount .. aside from other more restricted permissions and file system types that might not work well with locks.

if i had an easy answer to this, i would have already done it, but it is much harder than you would think at first approach.

Reply:
Well, of course there will always be a exception,where it won't work. The whole locking functionality is a hack anyway. But from my point of view, it's better to have it working on 95% systems, than not having this feature at all.

Maybe start with /tmp and see on which systems this will cause problems?
Does any system use polyinstantiation of /tmp by default?

Reply:
that isn't only issue, many users require specific directories on their remotes for different reasons, this violates 'remote_tmp' configurations, which are normally per user, that is why most of our focus on creating these lock files had to do with the target file and not a common temp dir.